spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.conformance.interpreter.fixtures
+
+import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.catalyst.encoders.RowEncoder
+import org.apache.spark.sql.execution.streaming.MemoryStream
+import org.apache.spark.sql.streaming.Trigger
+import org.scalatest.FunSuite
+import org.scalatest.mockito.MockitoSugar
+import za.co.absa.enceladus.conformance.{ConfCmdConfig, HyperConformance}
+import za.co.absa.enceladus.conformance.interpreter.FeatureSwitches
+import za.co.absa.enceladus.conformance.streaming.{InfoDateFactory, InfoDateLiteralFactory}
+import za.co.absa.enceladus.dao.MenasDAO
+import za.co.absa.enceladus.model.Dataset
+import za.co.absa.enceladus.utils.testUtils.SparkTestBase
+
+trait StreamingFixture extends FunSuite with SparkTestBase with MockitoSugar {
+  implicit val menasBaseUrls: List[String] = List()
+  implicit val cmd: ConfCmdConfig = ConfCmdConfig.apply(reportVersion = Some(1))
+  implicit val infoDateFactory: InfoDateFactory = new InfoDateLiteralFactory("2020-05-23")
+  private val interval = 10000
+
+  protected def testHyperConformance(input: DataFrame,
+                                     memoryTableName: String,
+                                     dataset: Dataset,
+                                     catalystWorkaround: Boolean = true)
+                                    (implicit menasDAO: MenasDAO): DataFrame = {
+    implicit val featureSwitches: FeatureSwitches = FeatureSwitches()
+      .setExperimentalMappingRuleEnabled(false)
+      .setCatalystWorkaroundEnabled(catalystWorkaround)
+      .setControlFrameworkEnabled(false)
+
+    val memoryStream = new MemoryStream[Row](1, spark.sqlContext)(RowEncoder(input.schema))
+    val hyperConformance = new HyperConformance()
+    val source: DataFrame = memoryStream.toDF()
+    val conformed: DataFrame = hyperConformance.applyConformanceTransformations(source, dataset)
+    val sink = conformed
+      .writeStream
+      .trigger(Trigger.ProcessingTime(interval))
+      .queryName(memoryTableName)
+      .outputMode("append")
+      .format("memory")
+      .start()
+
+    input.collect().foreach(e => memoryStream.addData(e))
+    sink.awaitTermination(interval)
```

Comment:
`awaitTermination(interval)` is probably a little bit dangerous. If the query is terminated just before the microbatch has been committed, then the whole microbatch will be aborted and there will be zero rows.
Also, you have to assume that your data will be processed within `interval` ms. 10s is probably enough, but it's kind of a magic value.
In this case `processAllAvailable` would be a good fit, because then you can be sure that all the rows will have been read, no matter to what value `interval` has been set to. `processAllAvailable` terminates as long as the data source doesn't emit values indefinitely which is the case here

Reply:
I agreee, but how can you use `processAllAvailable` on a test with the catalyst optimizer workaround disabled without having an unreasonable running time for testing?

Reply:
Ok, now I see your need for a timeout. Either you give up on testing with multiple micro-batches and use `Trigger.Once` or you could use a margin of safety, say `3 * interval`.
Also, if you really want to cover multiple micro-batches, you need to make sure that the `foreach` loop runs during multiple micro-batches.

Reply:
I'm slightly confused. We need to test 2 things:
1. The correctness of the results of streaming conformance. All workarounds should be turned on so such tests can finish in a reasonable time.
2. We need to test that workaround works by adding a test that hangs if the workaround is turned off. There is no particular time limit on the test, but should be some reasonable time.

Tests that test (1) should not rely on timing if possible. If `processAllAvailable` helps with it - great..
Tests that test (2) might rely on timing, but not necessarily. Currently in batch, if the workaround stops working the corresponding test will just hang forever. Not ideal, but we expect workaround to work. If it stops having an effect, it will be noticed by the developer when the test hangs.
Alternatively, a big timeout can be used to interrupt the test if it takes too long. But the value of the timeout should be set so that test passes on all envs the test could run, including Jenkins VMs.

The reason I care about this so much is that timings introduce non-determinism. Such tests could work perfectly locally on Mac, but start failing on Windows or on Linux build box.